File: 2021/day05/php/solution.php
<?php

function parseInput(): array {
    $inputPath = __DIR__ . '/../input.txt';
    $content = file_get_contents($inputPath);
    $lines = [];

    foreach (explode("\n", $content) as $line) {
        $line = trim($line);
        if (empty($line)) {
            continue;
        }
        $parts = explode(' -> ', $line);
        list($x1, $y1) = array_map('intval', explode(',', $parts[0]));
        list($x2, $y2) = array_map('intval', explode(',', $parts[1]));
        $lines[] = [$x1, $y1, $x2, $y2];
    }

    return $lines;
}

function sign(int $x): int {
    if ($x > 0) return 1;
    if ($x < 0) return -1;
    return 0;
}

function countOverlaps(array $lines, bool $includeDiagonals = false): int {
    $grid = [];

    foreach ($lines as $line) {
        list($x1, $y1, $x2, $y2) = $line;
        $dx = sign($x2 - $x1);
        $dy = sign($y2 - $y1);

        // Skip diagonals in part 1
        if (!$includeDiagonals && $dx !== 0 && $dy !== 0) {
            continue;
        }

        $x = $x1;
        $y = $y1;
        while (true) {
            $key = "$x,$y";
            if (!isset($grid[$key])) {
                $grid[$key] = 0;
            }
            $grid[$key]++;

            if ($x === $x2 && $y === $y2) {
                break;
            }
            $x += $dx;
            $y += $dy;
        }
    }

    $count = 0;
    foreach ($grid as $value) {
        if ($value >= 2) {
            $count++;
        }
    }
    return $count;
}

function part1(array $lines): int {
    return countOverlaps($lines, false);
}

function part2(array $lines): int {
    return countOverlaps($lines, true);
}

$lines = parseInput();
echo "Part 1: " . part1($lines) . "\n";
echo "Part 2: " . part2($lines) . "\n";

File: 2021/day05/rust/solution.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;

struct Line {
    x1: i32,
    y1: i32,
    x2: i32,
    y2: i32,
}

fn parse_input() -> Vec<Line> {
    let input_path = Path::new(file!()).parent().unwrap().join("../input.txt");
    let content = fs::read_to_string(&input_path).expect("Failed to read input file");

    content
        .lines()
        .filter(|line| !line.is_empty())
        .map(|line| {
            let parts: Vec<&str> = line.split(" -> ").collect();
            let start: Vec<i32> = parts[0].split(',').map(|x| x.parse().unwrap()).collect();
            let end: Vec<i32> = parts[1].split(',').map(|x| x.parse().unwrap()).collect();
            Line {
                x1: start[0],
                y1: start[1],
                x2: end[0],
                y2: end[1],
            }
        })
        .collect()
}

fn sign(x: i32) -> i32 {
    match x.cmp(&0) {
        std::cmp::Ordering::Greater => 1,
        std::cmp::Ordering::Less => -1,
        std::cmp::Ordering::Equal => 0,
    }
}

fn count_overlaps(lines: &[Line], include_diagonals: bool) -> usize {
    let mut grid: HashMap<(i32, i32), i32> = HashMap::new();

    for line in lines {
        let dx = sign(line.x2 - line.x1);
        let dy = sign(line.y2 - line.y1);

        // Skip diagonals in part 1
        if !include_diagonals && dx != 0 && dy != 0 {
            continue;
        }

        let mut x = line.x1;
        let mut y = line.y1;
        loop {
            *grid.entry((x, y)).or_insert(0) += 1;
            if x == line.x2 && y == line.y2 {
                break;
            }
            x += dx;
            y += dy;
        }
    }

    grid.values().filter(|&&v| v >= 2).count()
}

fn part1(lines: &[Line]) -> usize {
    count_overlaps(lines, false)
}

fn part2(lines: &[Line]) -> usize {
    count_overlaps(lines, true)
}

fn main() {
    let lines = parse_input();
    println!("Part 1: {}", part1(&lines));
    println!("Part 2: {}", part2(&lines));
}
